transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -0,0 +1,355 @@
+                      .setEventMillis(cursor.getLong(3))
+                      .setUptimeMillis(cursor.getLong(4))
+                      .setPayload(cursor.getBlob(5))
+                      .build();
+              events.add(PersistedEvent.create(id, backendName, event));
+            }
+          }
+
+          Map<Long, Set<Metadata>> metadataIndex = new HashMap<>();
+          StringBuilder whereClause = new StringBuilder("event_id IN (");
+          for (int i = 0; i < events.size(); i++) {
+            whereClause.append(events.get(i).getId());
+            if (i < events.size() - 1) {
+              whereClause.append(',');
+            }
+          }
+          whereClause.append(')');
+
+          try (Cursor cursor =
+              db.query(
+                  "event_metadata",
+                  new String[] {"event_id", "name", "value"},
+                  whereClause.toString(),
+                  null,
+                  null,
+                  null,
+                  null)) {
+            while (cursor.moveToNext()) {
+              long eventId = cursor.getLong(0);
+              Set<Metadata> currentSet = metadataIndex.get(eventId);
+              if (currentSet == null) {
+                currentSet = new HashSet<>();
+                metadataIndex.put(eventId, currentSet);
+              }
+              currentSet.add(new Metadata(cursor.getString(1), cursor.getString(2)));
+            }
+          }
+          ListIterator<PersistedEvent> iterator = events.listIterator();
+          while (iterator.hasNext()) {
+            PersistedEvent current = iterator.next();
+            if (!metadataIndex.containsKey(current.getId())) {
+              continue;
+            }
+            EventInternal.Builder newEvent = current.getEvent().toBuilder();
+
+            for (Metadata metadata : metadataIndex.get(current.getId())) {
+              newEvent.addMetadata(metadata.key, metadata.value);
+            }
+            iterator.set(
+                PersistedEvent.create(current.getId(), current.getBackendName(), newEvent.build()));
+          }
+          return events;
+        });
+  }
+
+  interface TransactionFn<T> {
+    T execute(SQLiteDatabase db);
+  }
+
+  private <T> T inTransaction(TransactionFn<T> function) {
+    SQLiteDatabase db = getDb();
+    db.beginTransaction();
+    try {
+      T result = function.execute(db);
+      db.setTransactionSuccessful();
+      return result;
+    } finally {
+      db.endTransaction();
+    }
+  }
+
+  private static class Metadata {
+    final String key;
+    final String value;
+
+    private Metadata(String key, String value) {
+      this.key = key;
+      this.value = value;
+    }
+  }
+
+  private static Priority toPriority(int value) {
+    if (value < 0 || value >= ALL_PRIORITIES.length) {
+      return Priority.DEFAULT;
+    }
+    return ALL_PRIORITIES[value];
+  }
+
+  private static class OpenHelper extends SQLiteOpenHelper {
```

Comment:
Your OpenHelper could be better set up to handle schema changes.

Consider instead structuring this as a series of migrations, and then run the migrations required to get from one version to the next. See [Firestore's implementation](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteSchema.java#L75) for ideas.

This also makes it easy to test that the migrations are happening correctly in downgrade/upgrade scenarios. See [SQLiteSchemaTest](https://github.com/firebase/firebase-android-sdk/blob/master/firebase-firestore/src/test/java/com/google/firebase/firestore/local/SQLiteSchemaTest.java)

Reply:
You can refactor when you need it for sure.

The problematic case is when you do ship schema v2, you need to ensure that your migrations are idempotent after a user downgrades to 1 and then re-upgrades to 2. Firestore didn't do this initially and so we ended up failing trying to create tables that already existed.

If you're not going to use migrations now, make sure to leave a monster TODO indicating SCHEMA_VERSION cannot change until you can handle re-upgrades after downgrades.

Reply:
Fair enough. Will document in properly